File: ManagerUtils/PlotUtils/src/PlotObj.cc
/*******************************************************************************
*
*  Filename    : PlotObj.cc
*  Description : Plotting Object declaration
*  Author      : Yi-Mu "Enoch" Chen [ ensc@hep1.phys.ntu.edu.tw ]
*
*******************************************************************************/

#include "/wk_cms2/shihlin0314/CMSSW_8_0_29/src/ManagerUtils/PlotUtils/interface/Constants.hpp"
#include "TCanvas.h"
#include "TGraphAsymmErrors.h"
#include "TH1D.h"
#include "TLegend.h"
#include "TPad.h"
#include "TPaveText.h"

namespace mgr{
    /*******************************************************************************
    *   Pad settings functions
    *******************************************************************************/
    static const float TOP_BOTTOM_SEP = 0.3;

    TCanvas*
    NewCanvas()
    {
        TCanvas* c = new TCanvas( "c", "c", DEFAULT_CANVAS_WIDTH, DEFAULT_CANVAS_HEIGHT );
        c->SetTicks( 1, 1 );
        return c;
    }

    /******************************************************************************/

    TPad*
    NewTopPad()
    {
        TPad* pad = new TPad( "toppad", "", 0, TOP_BOTTOM_SEP, 1., 1.0 );
        pad->SetTicks( 1, 1 );
        pad->SetBottomMargin( 0.05 );
        pad->SetLeftMargin( PLOT_X_MIN );
        pad->SetRightMargin( 1 - PLOT_X_MAX );
        pad->SetTopMargin( ( 1 - PLOT_Y_MAX ) / ( 1 - TOP_BOTTOM_SEP ) );
        return pad;
    }

    /******************************************************************************/

    TPad*
    NewBottomPad()
    {
        TPad* pad = new TPad( "botpad", "", 0, 0.00, 1, TOP_BOTTOM_SEP );
        pad->SetTicks( 1, 1 );
        pad->SetTopMargin( 0.025 );
        pad->SetLeftMargin( PLOT_X_MIN );
        pad->SetRightMargin( 1 - PLOT_X_MAX );
        pad->SetBottomMargin( ( PLOT_Y_MIN ) / ( TOP_BOTTOM_SEP ) );
        return pad;
    }

    /******************************************************************************/

    void
    SetSinglePad( TVirtualPad* pad )
    {
	pad->SetTicks( 1, 1);        
	pad->SetLeftMargin( PLOT_X_MIN );
        //pad->SetLeftMargin( 0.2);
	pad->SetRightMargin( 1 - PLOT_X_MAX );
        pad->SetBottomMargin( PLOT_Y_MIN );
        pad->SetTopMargin( 1 - PLOT_Y_MAX );
    }

    /******************************************************************************/

    void
    SetSinglePadWithPalette( TVirtualPad* pad )
    {
        pad->SetLeftMargin( PLOT_X_MIN );
        pad->SetBottomMargin( PLOT_Y_MIN );
        pad->SetTopMargin( 1 - PLOT_Y_MAX );
    }

    /*******************************************************************************
    *   New Legend settings
    *******************************************************************************/
    TLegend*
    NewLegend( const float x_min,
               const float y_min,
               const float x_max,
               const float y_max )
    {
        TLegend* ans = new TLegend( x_min, y_min, x_max, y_max );
        // Setting up default values
        ans->SetBorderSize( 0 );
        ans->SetTextFont( FONT_TYPE );
        ans->SetTextSize( TEXT_FONT_SIZE );
        ans->SetFillColorAlpha( 0, 0 );
        return ans;
    }

    /******************************************************************************/

    TPaveText*
    NewTextBox( const float x_min,
                const float y_min,
                const float x_max,
                const float y_max )
    {
        TPaveText* ans = new TPaveText( x_min, y_min, x_max, y_max, "NDC" );
        ans->SetTextFont( FONT_TYPE );
        ans->SetTextSize( TEXT_FONT_SIZE );
        ans->SetFillColor( kWhite );
        return ans;
    }

    /*******************************************************************************
    *   TH1D division functions
    *******************************************************************************/
    TH1D*
    DivideHist(
        TH1D*        num,
        TH1D*        den,
        const double cen
        )
    {
        TH1D* ans = (TH1D*)( num->Clone() );
        ans->Divide( den );

        for( int i = 0; i < num->GetNcells(); ++i ){
            if( num->GetBinContent( i ) == 0 && den->GetBinContent( i ) == 0 ){
                ans->SetBinContent( i, cen );
            }
        }

        return ans;
    }

    /*******************************************************************************
    *   TGraph division functions
    *******************************************************************************/
    TGraph*
    DividedGraphSimple(
        TGraph* num,
        TGraph* den
        )
    {
        TGraph* ans = new TGraph( num->GetN() );

        for( int i = 0; i < num->GetN(); ++i ){
            const double origx = num->GetX()[ i ];
            const double origy = num->GetY()[ i ];
            const double deny  = den->Eval( origx );

            if( fabs( deny ) > 0.0000001 ){
                ans->SetPoint( i, origx, origy / deny );
            }
            else{
                ans->SetPoint( i, origx, 1 );
            }
        }

        ans->SetTitle( "" );
        return ans;
    }

    /******************************************************************************/

    TGraphAsymmErrors*
    DividedGraph(
        TGraphAsymmErrors* num,
        TGraph*            den
        )
    {
        TGraphAsymmErrors* ans = new TGraphAsymmErrors( num->GetN() );

        for( int i = 0; i < num->GetN(); ++i ){
            const double origx  = num->GetX()[ i ];
            const double origy  = num->GetY()[ i ];
            const double xerrlo = num->GetErrorXlow( i );
            const double xerrhi = num->GetErrorXhigh( i );
            const double yerrlo = num->GetErrorYlow( i );
            const double yerrhi = num->GetErrorYhigh( i );
            const double deny   = den->Eval( origx );

            if( deny > 0.0000001 ){
                ans->SetPoint( i, origx, origy / deny );
                ans->SetPointError(
                    i,
                    xerrlo,
                    xerrhi,
                    yerrlo / deny,
                    yerrhi / deny
                    );
            }
            else{
                ans->SetPoint( i, origx, 1 );
                ans->SetPointError(
                    i,
                    xerrlo,
                    xerrhi,
                    0,
                    0
                    );
            }
        }

        ans->SetTitle( "" );
        return ans;
    }

    /******************************************************************************/

    TGraphAsymmErrors*
    PullGraph(
        TGraph*            numgraph,
        TGraphAsymmErrors* dengraph
        )
    {
        TGraphAsymmErrors* ans = new TGraphAsymmErrors( dengraph->GetN() );

        for( int i = 0; i < dengraph->GetN(); ++i ){
            const double x    = dengraph->GetX()[ i ];
            const double y    = dengraph->GetY()[ i ];
            const double numy = numgraph->Eval( x );
            const double pull = y == 0 ? 0 :
                                numy > y ? ( numy - y ) / dengraph->GetErrorYhigh( i ) :
                                ( numy - y ) / dengraph->GetErrorYlow( i );
            ans->SetPoint( i, x, pull );
            ans->SetPointError(
                i,
                dengraph->GetErrorX( i ),
                dengraph->GetErrorX( i ),
                0, 0// No y error
                );
        }

        return ans;
    }
}

File: ManagerUtils/PlotUtils/python/PlotObj.py
from ManagerUtils.PlotUtils.Constants import *
import ROOT
# /*******************************************************************************
# *   Pad settings functions
# *******************************************************************************/
def NewCanvas( obj="c" ):
    ROOT.gROOT.SetBatch(ROOT.kTRUE)
    c = ROOT.TCanvas( obj, obj, DEFAULT_CANVAS_WIDTH, DEFAULT_CANVAS_HEIGHT )
    return c

def NewTopPad():
    pad = ROOT.TPad( "toppad", "", 0., TOP_BOTTOM_SEP, 1., 1. )
    pad.SetTicks( 1, 1 )
    pad.SetBottomMargin( 0.05 )
    pad.SetLeftMargin( PLOT_X_MIN )
    pad.SetRightMargin( 1 - PLOT_X_MAX )
    pad.SetTopMargin( ( 1 - PLOT_Y_MAX ) / ( 1 - TOP_BOTTOM_SEP ) )
    return pad

def NewBottomPad():
    pad = ROOT.TPad( "botpad", "", 0., 0.0, 1., TOP_BOTTOM_SEP )
    pad.SetTicks( 1, 1 )
    pad.SetTopMargin( 0.025 )
    pad.SetLeftMargin( PLOT_X_MIN )
    pad.SetRightMargin( 1 - PLOT_X_MAX )
    pad.SetBottomMargin( ( PLOT_Y_MIN ) / ( TOP_BOTTOM_SEP ) )
    return pad

def SetSinglePad( c ):
    c.SetTicks( 1, 1)
    c.SetLeftMargin( PLOT_X_MIN )
    c.SetRightMargin( 1 - PLOT_X_MAX )
    c.SetBottomMargin( PLOT_Y_MIN )
    c.SetTopMargin( 1 - PLOT_Y_MAX )

def SetSinglePadWithPalette( c ): 
    c.SetTicks( 1, 1)
    c.SetLeftMargin( PLOT_X_MIN )
    c.SetBottomMargin( PLOT_Y_MIN )
    c.SetTopMargin( 1 - PLOT_Y_MAX )

# /*******************************************************************************
# *   New Legend settings
# *******************************************************************************/
def NewLegend( xmin, ymin, xmax, ymax ):
    leg = ROOT.TLegend( xmin, ymin, xmax, ymax )
    leg.SetBorderSize( 0 )
    leg.SetTextFont( FONT_TYPE )
    leg.SetTextSize( TEXT_FONT_SIZE )
    leg.SetFillColorAlpha( 0, 0 )
    leg.SetLineColor( ROOT.kWhite )
    return leg

def NewTextBox( xmin, ymin, xmax, ymax ):
    text = ROOT.TPaveText( xmin, ymin, xmax, ymax, "NDC" )
    text.SetTextFont( FONT_TYPE )
    text.SetTextSize( TEXT_FONT_SIZE )
    text.SetFillColor( ROOT.kWhite )
    return text

def DrawEntryLeft( entry ):
    tl = ROOT.TLatex()
    tl.SetNDC( True )
    tl.SetTextFont( FONT_TYPE )
    tl.SetTextSize( TEXT_FONT_SIZE )
    tl.SetTextAlign( BOTTOM_LEFT )
    tl.DrawLatex( PLOT_X_MIN, PLOT_Y_MAX + TEXT_MARGIN / 2, entry )

def DrawEntryRight( entry ):
    tl = ROOT.TLatex()
    tl.SetNDC( True )
    tl.SetTextFont( FONT_TYPE )
    tl.SetTextSize( TEXT_FONT_SIZE )
    tl.SetTextAlign( BOTTOM_RIGHT )
    tl.DrawLatex( PLOT_X_MAX, PLOT_Y_MAX + ( TEXT_MARGIN / 2 ), entry )

# /*******************************************************************************
# *    Plot CMS label
# *******************************************************************************/
def DrawCMSLabel( tag ):
    tl = ROOT.TLatex()
    tl.SetNDC( True )
    tl.SetTextFont( FONT_TYPE )
    tl.SetTextSize( AXIS_TITLE_FONT_SIZE )
    tl.SetTextAlign( TOP_LEFT )
    tl.DrawLatex( PLOT_X_MIN + TEXT_MARGIN, PLOT_Y_MAX - TEXT_MARGIN, "#bf{CMS}" )

    if tag == PRELIMINARY:
        tl.DrawLatex( PLOT_X_MIN + TEXT_MARGIN, PLOT_Y_MAX - TEXT_MARGIN - LINE_HEIGHT, "#it{Preliminary}" )
    
    elif tag == SIMULATION:
        tl.DrawLatex( PLOT_X_MIN + TEXT_MARGIN, PLOT_Y_MAX - TEXT_MARGIN - LINE_HEIGHT, "#it{Simulation}" )

def DrawCMSLabelOuter( tag ):
    tl = ROOT.TLatex()
    tl.SetNDC( True )
    tl.SetTextFont( FONT_TYPE )
    tl.SetTextSize( AXIS_TITLE_FONT_SIZE )
    tl.SetTextAlign( BOTTOM_LEFT )
    text = "#bf{CMS} "

    if tag == PRELIMINARY:
        text += "#it{Preliminary}"
    elif tag == SIMULATION:
        text += "#it{Simulation}"

    tl.DrawLatex( PLOT_X_MIN, PLOT_Y_MAX + TEXT_MARGIN / 2, text )

def DrawLuminosity( lumi ):
    tl = ROOT.TLatex()
    tl.SetNDC( True )
    tl.SetTextFont( FONT_TYPE )
    tl.SetTextSize( AXIS_TITLE_FONT_SIZE )
    buffer = "{0:.1f}fb^{{-1}} ({1} TeV)".format( lumi / 1000., 13 )
    tl.SetTextAlign( BOTTOM_RIGHT )
    tl.DrawLatex( PLOT_X_MAX, PLOT_Y_MAX + ( TEXT_MARGIN / 2 ), buffer )
